src/lib/operators/set_union.cpp

```diff
@@ -0,0 +1,371 @@
+}
+
+const std::string SetUnion::name() const { return "SetUnion"; }
+
+const std::string SetUnion::description() const { return "SetUnion"; }
+
+std::shared_ptr<const Table> SetUnion::_on_execute() {
+  const auto early_result = _analyze_input();
+  if (early_result) {
+    return early_result;
+  }
+
+  /**
+   * For each input, create a ReferenceMatrix
+   */
+  const auto build_reference_matrix = [&](const auto& input_table, auto& reference_matrix) {
+    reference_matrix.resize(_column_segment_begins.size());
+    for (auto& pos_list : reference_matrix) {
+      pos_list.reserve(input_table->row_count());
+    }
+
+    for (auto chunk_id = ChunkID{0}; chunk_id < input_table->chunk_count(); ++chunk_id) {
+      const auto& chunk = input_table->get_chunk(ChunkID{chunk_id});
+
+      for (size_t segment_id = 0; segment_id < _column_segment_begins.size(); ++segment_id) {
+        const auto column_id = _column_segment_begins[segment_id];
+        const auto column = chunk.get_column(column_id);
+        const auto ref_column = std::dynamic_pointer_cast<const ReferenceColumn>(column);
+
+        auto& out_pos_list = reference_matrix[segment_id];
+        auto in_pos_list = ref_column->pos_list();
+        std::copy(in_pos_list->begin(), in_pos_list->end(), std::back_inserter(out_pos_list));
+      }
+    }
+  };
+
+  ReferenceMatrix reference_matrix_left;
+  ReferenceMatrix reference_matrix_right;
+
+  build_reference_matrix(_input_table_left(), reference_matrix_left);
+  build_reference_matrix(_input_table_right(), reference_matrix_right);
+
+  /**
+   * Init the virtual pos lists
+   */
+  VirtualPosList virtual_pos_list_left(_input_table_left()->row_count(), 0u);
+  std::iota(virtual_pos_list_left.begin(), virtual_pos_list_left.end(), 0u);
+  VirtualPosList virtual_pos_list_right(_input_table_right()->row_count(), 0u);
+  std::iota(virtual_pos_list_right.begin(), virtual_pos_list_right.end(), 0u);
+
+  /**
+   * Sort the virtual pos lists, so we can merge them
+   */
+  std::sort(virtual_pos_list_left.begin(), virtual_pos_list_left.end(),
+            VirtualPosListCmpContext{reference_matrix_left});
+  std::sort(virtual_pos_list_right.begin(), virtual_pos_list_right.end(),
+            VirtualPosListCmpContext{reference_matrix_right});
+
+  /**
+   * Build result table
+   */
+  auto left_idx = size_t{0};
+  auto right_idx = size_t{0};
+  const auto num_left = virtual_pos_list_left.size();
+  const auto num_right = virtual_pos_list_right.size();
+
+  // Somewhat random way to decide on a chunk size.
+  const auto out_chunk_size = std::max(_input_table_left()->chunk_size(), _input_table_right()->chunk_size());
+
+  auto out_table = Table::create_with_layout_from(_input_table_left(), out_chunk_size);
+
+  std::vector<std::shared_ptr<PosList>> pos_lists(reference_matrix_left.size());
+  std::generate(pos_lists.begin(), pos_lists.end(), [&] { return std::make_shared<PosList>(); });
+
+  // Adds one row to the pos_lists currently being filled
```

Comment:
"Adds a row"? I first understood "one" to be one specifically.

Reply:
explained it a bit better